Generalize proto "gluon_wired.sh"

lemoer opened the pull request:
These are two small patches to allow using the `gluon_wired.sh` proto for the `mesh_vpn` interface.

This works for layer 2 and layer 3 vpns:
- In case of a layer 2 vpn, this just adds the `gluon_mesh.sh` proto as before.
- In case of a layer 3 vpn, creates an intermediate vxlan interface on top of the interface before applying the `gluon_mesh.sh` proto.

Furthermore an additional setting `vxpeer6addr` is added to the gluon_wired.sh proto, which is only used, if a vxlan layer is added. It can be used to control the peer address of vxlan.

This is preparatory to support wireguard in a clean way.

mweinelt commented on package/gluon-core/files/lib/netifd/proto/gluon_wired.sh:
```diff
@@ -7,15 +7,29 @@ init_proto "$@"
 
+is_layer3_device () {
+        local addrlen="$(cat "/sys/class/net/$1/addr_len")"
+        test "$addrlen" -eq 0
```
Only L2 devices have MAC addresses, hence an `addr_len` of 0 bytes indicates a L3 device.

```
What:		/sys/class/net/<iface>/addr_len
Date:		April 2005
KernelVersion:	2.6.12
Contact:	netdev@vger.kernel.org
Description:
		Indicates the hardware address size in bytes.
		Values vary based on the lower-level protocol used by the
		interface (Ethernet, FDDI, ATM, IEEE 802.15.4...). See
		include/uapi/linux/if_*.h for actual values.
```

lemoer replied:
That's correct.

mweinelt commented on package/gluon-core/files/lib/netifd/proto/gluon_wired.sh:
```diff
@@ -28,10 +42,22 @@ proto_gluon_wired_setup() {
+        [ -z "$is_tunnel" ] && is_tunnel=0
+        [ -z "$vxpeer6addr" ] && vxpeer6addr='ff02::15c'
+
+        if [ "$is_tunnel" -eq 1 ] && is_layer3_device "$ifname"; then
```
Can you explain what `is_tunnel` represents? I'm not sure I follow the reasoning here.

lemoer replied:
It is possible that `vxlan` for *mesh_on_lan* and *mesh_on_wan* is disabled in the site/domain.

However for vpn we want to have vxlan if the interface is a layer 3 vpn and disabled if we have a layer 2 vpn. There is no correlation to the `vxlan` setting from site.

As I was not able to get the information whether an interface is a tunnel from somewhere else, I introduced the proto variable `is_tunnel`.

mweinelt replied:
I wonder if we can get read `IFF_TUN` and `IFF_TAP` from `/sys/class/net/$ifname/tun_flags`.

https://elixir.bootlin.com/linux/v4.14.213/source/include/uapi/linux/if_tun.h#L62

lemoer replied:
This does not seem to exist for wireguard devices.

mweinelt replied:
No sensible way I see to reliably detect a tunnel interface in a consistent fashion.

NeoRaider replied:
In libgluonutil we have this code: https://github.com/freifunk-gluon/gluon/blob/master/package/libgluonutil/src/libgluonutil.c#L156-L202

lemoer replied:
Not necessary anymore.

mweinelt commented on package/gluon-core/files/lib/netifd/proto/gluon_wired.sh:
```diff
@@ -7,15 +7,29 @@ init_proto "$@"
 interface_linklocal() {
+        if is_layer3_device "$1"; then
+                # this is probably a layer 3 device
+                ubus call "network.interface.$1" status  |  \
```
Where is the status command coming from?

```
# ubus -v list network
'network' @4aeb533b
	"restart":{}
	"reload":{}
	"add_host_route":{"target":"String","v6":"Boolean","interface":"String"}
	"get_proto_handlers":{}
	"add_dynamic":{"name":"String"}
```

mweinelt commented on package/gluon-core/files/lib/netifd/proto/gluon_wired.sh:
```diff
@@ -7,15 +7,29 @@ init_proto "$@"
+
 # shellcheck disable=SC2086
 interface_linklocal() {
+        if is_layer3_device "$1"; then
```
What converts `mesh-vpn` (from https://github.com/freifunk-gluon/gluon/pull/2163/files#diff-44dd4d020d9ae0de557dc06405de0e7ce445eeb893d001cf1914a932f8bbedf6R11) to `mesh_vpn` needed for the ubus call?

lemoer replied:
Netifd does that. `$ifname` (fetched from `$2`) in this script is **not** `ifname` from the uci section. `$ifname` is already resolved to `mesh_vpn` in this script.

I really dislike the naming conventions in netifd. It's just unnecessarily confusing. Scripts seem to name this argument `$2` differently:
```
# awk '/^proto_.*_setup/ { P=1; printf $1; } /^}/ {if (P) printf "\n"; P=0} /.*\$2.*/ { if(P) printf $0 }' /lib/netifd/proto/*
proto_batadv_setup()
proto_batadv_hardif_setup()	local iface="$2"
proto_batadv_vlan_setup()	local iface="$2"
proto_dhcp_setup()	local iface="$2"
proto_dhcpv6_setup()	local iface="$2"
proto_gluon_bat0_setup()
proto_gluon_mesh_setup()	export IFNAME="$2"
proto_gluon_wired_setup()        local ifname="$2"
proto_vxlan_setup()
proto_vxlan6_setup()
```

Naming it `iface` (like e.g. *proto_batadv_hardif_setup()*) seems to be better than naming it `ifname`. However still *proto_batadv_setup()* defines `iface="$config"` which means that `$iface=$1`:

```
root@lemoer-wgtest:/lib/netifd/proto# awk '/^proto_.*_setup/ { P=1; printf $1; } /^}/ {if (P) printf "\n"; P=0} /.*iface=.*/ { if(P) printf $0 }' /lib/netifd/proto/*
proto_batadv_setup()	local iface="$config"
proto_batadv_hardif_setup()	local iface="$2"
proto_batadv_vlan_setup()	local iface="$2"
proto_dhcp_setup()	local iface="$2"
proto_dhcpv6_setup()	local iface="$2"
proto_gluon_bat0_setup()
proto_gluon_mesh_setup()
proto_gluon_wired_setup()
proto_vxlan_setup()
proto_vxlan6_setup()
```

It's a mess in my oppinion. I don't know why netifd doesn't introduce something like "lower interface" or so. But maybe I am just not getting the concept...

mweinelt replied:
Yet the call in L19/L33 only works when `$1` is `mesh-vpn`.

```sh
# ubus call network.device status '{"name": "mesh_vpn"}'
Command failed: Invalid argument
```

vs.

```sh
# ubus call network.device status '{"name": "mesh-vpn"}'
{
	"external": false,
	"present": true,
	"type": "Network device",
	"up": true,
	"carrier": true,
	"link-advertising": [
		
	],
	"link-partner-advertising": [
		
	],
	"link-supported": [
		
	],
	"speed": "10F",
	"autoneg": false,
	"mtu": 1312,
	"mtu6": 1312,
	"macaddr": "42:b6:0e:6d:16:c7",
	"txqueuelen": 1000,
	"ipv6": true,
	"promisc": false,
	"rpfilter": 0,
	"acceptlocal": false,
	"igmpversion": 0,
	"mldversion": 0,
	"neigh4reachabletime": 30000,
	"neigh6reachabletime": 30000,
	"neigh4gcstaletime": 60,
	"neigh6gcstaletime": 60,
	"neigh4locktime": 100,
	"dadtransmits": 1,
	"multicast": true,
	"sendredirects": true,
	"statistics": {
		"collisions": 0,
		"rx_frame_errors": 0,
		"tx_compressed": 0,
		"multicast": 0,
		"rx_length_errors": 0,
		"tx_dropped": 0,
		"rx_bytes": 0,
		"rx_missed_errors": 0,
		"tx_errors": 0,
		"rx_compressed": 0,
		"rx_over_errors": 0,
		"tx_fifo_errors": 0,
		"rx_crc_errors": 0,
		"rx_packets": 0,
		"tx_heartbeat_errors": 0,
		"rx_dropped": 0,
		"tx_aborted_errors": 0,
		"tx_packets": 10560199,
		"rx_errors": 0,
		"tx_bytes": 1723304086,
		"tx_window_errors": 0,
		"rx_fifo_errors": 0,
		"tx_carrier_errors": 0
	}
}
```

lemoer replied:
Ok, well you are right, I am wrong. `$2` to `proto_*_setup()` is acutally a device name, not an interface name (speaking in netifd terminology). So we have a problem here.

lemoer replied:
```
ubus call network.interface dump | \
   jsonfilter -e "@.interface[@.l3_device='$1']['ipv6-address'][*].address" | \
   grep -e '^fe[89ab]' | head -n 1
```

lemoer replied:
Applied this.

lemoer commented:
I rebased and removed the `proto_add_host_dependency` from *gluon_wired* again. It is not necessary, even worse, it has bricked things.

This PR is now tested with mesh_on_wan and with fastd.

NeoRaider commented:
Extending `gluon_wired` seems to add some complexity. Maybe we want to introduce a separate `gluon_vpn` proto? Both protos could source a file with common code though.

I might be overlooking something, but I don't understand why the fastd config is modified by this PR. Can't that just use `gluon_mesh` as before, and we only use a different proto for L3 VPNs?

lemoer commented:
I experimented a bit and with introducing new proto. The drawback of this approach is that we would have to complicate the firewall rule generation, as it is not that simple anymore to select all interfaces with proto `gluon_wired` into the zone `gluon_wired`. We would have to rename the zone and add interfaces matching the new proto aswell, or we would have to duplicate the section.

In my understanding, `gluon_wired` is just a proto, which *can* add a vxlan layer between `gluon_mesh` and the actual interface. It makes sense that this would also work for layer 3 devices.

However, some of the complexity is added due to the fact that the decision if a vxlan layer is added in between, is dependent to the interface. I would suggest therefore to move this decision out of the proto by exposing a boolean config parameter `add_vxlan_layer` from the proto.

Regarding fastd: I just modified the general *mesh_vpn* proto as the section was in the *gluon-mesh-vpn-core* package and `gluon_wired` supports leaving out the vxlan layer. That was the easiest option without the need to move the *mesh_vpn* proto into the other packages. I can differentiate here, if that's desired.

lemoer commented:
I implemented the suggested changes.

I kept changing the proto of the vpn interface to `gluon_wired`, as I currently see no benefit in staying with `gluon_mesh` here, if we would want to introduce `gluon_wired` for wireguard devices anyways and `gluon_wired` supports omitting the creation of the vxlan layer.

lemoer commented:
I just pushed a new version of this patch. I generalized it without adding too much extra glue. It's ready for review (and afterwards merge).

NeoRaider commented on package/gluon-core/files/lib/netifd/proto/gluon_wired.sh:
```diff
@@ -7,15 +7,29 @@ init_proto "$@"
 proto_gluon_wired_init_config() {
         proto_config_add_boolean transitive
         proto_config_add_int index
+        proto_config_add_boolean add_vxlan_layer
```
Please move the addition and handling of this option to a separate commit. I think a shorter name like `vxlan` would look better.

The option would probably not exist for long, as non-VXLAN setups should be phased out eventually (but we don't have a schedule for that yet - will depend on the patches mentioned in #2155 I guess).

NeoRaider commented on package/gluon-core/files/lib/netifd/proto/gluon_wired.sh:
```diff
@@ -7,15 +7,29 @@ init_proto "$@"
+        if is_layer3_device "$1"; then
+                ubus call network.interface dump | \
+                        jsonfilter -e "@.interface[@.l3_device='$1']['ipv6-address'][*].address" | \
+                        grep -e '^fe[89ab]' | head -n 1
```
Hmm, I assume this is about L3 VPN tunnels like Wireguard? How will link-local addresses be assigned for that?

Maybe we should enforce the configuration of an explicit `vxip6addr` in addition to `vxpeer6addr` in such cases, and only support L2 interfaces when no local address is configured.

lemoer replied:
It's dynamically added via `proto_add_ipv6_address` in a netifd script.

lemoer replied:
Please let me know when you have decided if you want to introduce `vxip6addr` instead.

NeoRaider replied:
I think using the LL address of the underlying interface is fine for now, but the code will need error handling to make the setup fail when no such address exists.

NeoRaider commented on package/gluon-core/files/lib/netifd/proto/gluon_wired.sh:
```diff
@@ -28,24 +42,31 @@ proto_gluon_wired_setup() {
+                proto_notify_error "$config" "MISSING_OPTION_ADD_VXLAN_LAYER"
+                proto_block_restart "$config"
+                exit
+        fi
```
No need to emit an error here, we can just let the option default to 1.

NeoRaider commented on package/gluon-core/files/lib/netifd/proto/gluon_wired.sh:
```diff
@@ -56,6 +77,7 @@ proto_gluon_wired_setup() {
         json_init
         json_add_string name "${config}_mesh"
         json_add_string ifname "@${meshif}"
+        [ -n "$index" ] && json_add_string macaddr "$(lua -e "print(require('gluon.util').generate_mac($index))")"
```
Remove this, it should be meaningless in VXLAN mode (as the MAC address is set above), and incorrect otherwise (will modify the MAC address of an underlying interface, but we already set all MAC addresses as intended).

NeoRaider reviewed (changes requested):
I'm not sure how long-lived this solution will be. I'd like to get rid of the static index assignment in Gluon's network setup eventually, but I don't have a design for a replacement yet.

lemoer commented:
Just updated the PR. All changes suggested by @NeoRaider are now implemented.

I implemented the last change as extra commit to make changes clear. This means that the commits should be squashed during the merge.

AiyionPrime commented:
@mweinelt or @NeoRaider would one of you like to review again?

blocktrron commented on package/gluon-core/luasrc/lib/gluon/upgrade/220-interface-lan:
```diff
@@ -15,6 +15,7 @@ uci:section('network', 'interface', 'mesh_lan', {
 	igmp_snooping = false,
 	proto         = 'gluon_wired',
 	index         = 4,
+	vxlan = site.mesh.vxlan(true),
```
indentation of equals.

AiyionPrime replied:
resolved in 8fdc8e70740e38fd6c1946fa9a9fd61d0bdfceed

mweinelt commented on package/gluon-core/files/lib/netifd/proto/gluon_wired.sh:
```diff
@@ -8,15 +8,28 @@ proto_gluon_wired_init_config() {
+        if is_layer3_device "$1"; then
+                ubus call network.interface dump | \
+                        jsonfilter -e "@.interface[@.l3_device='$1']['ipv6-address'][*].address" | \
+                        grep -e '^fe[89ab]' | head -n 1
```
What does this pattern match? Why `9ab`?

mweinelt reviewed (approved):
Looks pretty solid to me

rotanid commented:
two approvals here, could be merged, can you please squash the indention fix commit @AiyionPrime @lemoer ?

lemoer commented:
@rotanid thx, done

blocktrron commented:
Just so this is noted somewhere - custom created mesh interface will require to have a vxlan property assigned to them.

blocktrron commented on package/gluon-core/files/lib/netifd/proto/gluon_wired.sh:
```diff
@@ -28,24 +46,27 @@ proto_gluon_wired_setup() {
-        local transitive index
-        json_get_vars transitive index
+        # default args
+        [ -z "$vxlan" ] && vxlan=1
```
maybe set the old default value using "$(lua -e 'print(require("gluon.site").mesh.vxlan(true))')"? This way it does not break custom interfaces w/o vxlan on sysupgrade.

lemoer replied:
This was never officially supported. Of course, we could do this and (very few) people would benefit. But I think we should not start introducing migration code for unsupported configuration ways.

rotanid commented:
> Just so this is noted somewhere - custom created mesh interface will require to have a vxlan property assigned to them.

so we need this in docs and/or release notes, does it break existing stuff?

blocktrron commented:
It's breaking custom modifications which can break on any upgrade anyway.

So release notes (if at all) should be fine.

lemoer commented:
Thanks :)

rotanid commented:
> It's breaking custom modifications which can break on any upgrade anyway.
> 
> So release notes (if at all) should be fine.

if we know something, we should warn ;)
